File: ai/pong_ai.py
class PongAI:
    def __init__(self, difficulty='medium'):
        self.difficulty = difficulty
        self.paddle_height = 100
        self.paddle_y = 250
        self.target_y = 250
        self.ball_prediction = None
        self.last_update_time = 0
        self.ball_history = []
        self.max_history = 5
        self.field_height = 500

        # Difficulty settings
        if difficulty == 'easy':
            self.reaction_time = 1.0
            self.error_margin = 40
            self.prediction_skill = 0.6
            self.max_speed = 3
        elif difficulty == 'medium':
            self.reaction_time = 0.7
            self.error_margin = 20
            self.prediction_skill = 0.8
            self.max_speed = 5
        else:  # Hard
            self.reaction_time = 0.5
            self.error_margin = 10
            self.prediction_skill = 0.95
            self.max_speed = 7

    def update(self, ball_x, ball_y, ball_dx, ball_dy, current_time):
        """Update AI with current ball position and calculate next move"""

        # Only update once per second to simulate human-like reaction time
        if current_time - self.last_update_time < self.reaction_time:
            return self.paddle_y

        self.last_update_time = current_time

        # Add ball position to history
        self.ball_history.append((ball_x, ball_y, ball_dx, ball_dy))
        if len(self.ball_history) > self.max_history:
            self.ball_history.pop(0)

        # Predict where ball will be when it reaches AI's side
        if ball_dx > 0:  # Ball is moving toward AI
            self._predict_ball_position(ball_x, ball_y, ball_dx, ball_dy)

        # Move paddle toward predicted position with some 'skill' factor
        if self.ball_prediction is not None:
            target_y = self.ball_prediction

            # Add some error based on difficulty
            import random
            error = random.uniform(-self.error_margin, self.error_margin)
            target_y += error

            # Clamp to field boundaries
            target_y = max(self.paddle_height / 2, min(self.field_height - self.paddle_height / 2, target_y))

            self.target_y = target_y

        # Move toward target at limited speed
        if abs(self.target_y - self.paddle_y) > self.max_speed:
            if self.target_y > self.paddle_y:
                self.paddle_y += self.max_speed
            else:
                self.paddle_y -= self.max_speed
        else:
            self.paddle_y = self.target_y

        return self.paddle_y

    def _predict_ball_position(self, ball_x, ball_y, ball_dx, ball_dy):
        """Predict where the ball will be when it reaches the AI's paddle"""

        # Simple prediction logic
        if ball_dx <= 0:
            return  # Ball moving away

        # Calculate time for ball to reach right side
        # This is a simplified calculation and doesn't account for bounces
        right_side_x = 800  # Assuming field width is 800
        time_to_reach = (right_side_x - ball_x) / ball_dx

        # Calculate y position at that time
        predicted_y = ball_y + (ball_dy * time_to_reach)

        # Account for bounces
        while predicted_y < 0 or predicted_y > self.field_height:
            if predicted_y < 0:
                predicted_y = -predicted_y
            elif predicted_y > self.field_height:
                predicted_y = 2 * self.field_height - predicted_y

        # Apply skill factor - higher prediction_skill means more accurate prediction
        if self.prediction_skill < 1.0:
            import random
            skill_error = random.uniform(0, 1.0 - self.prediction_skill) * self.field_height
            if random.random() > 0.5:
                skill_error = -skill_error
            predicted_y += skill_error

        self.ball_prediction = predicted_y

File: static/js/pong-ai.js
/**
 * File: static/js/pong-ai.js
 * Mô tả: Class quản lý AI cho trò chơi Pong
 */

class PongAI {
    constructor(options = {}) {
        // Các thiết lập độ khó
        this.difficulty = options.difficulty || 'medium';

        // Kích thước và vị trí paddle
        this.paddleHeight = options.paddleHeight || 100;
        this.paddleY = options.initialY || 250;
        this.fieldHeight = options.fieldHeight || 500;

        // Lưu trữ thông tin bóng để dự đoán
        this.ballHistory = [];
        this.maxHistory = 5;

        // Biến dự đoán
        this.targetY = this.paddleY;
        this.predictedY = null;

        // Thời gian
        this.lastUpdateTime = 0;

        // Thiết lập thông số dựa trên độ khó
        this.setupDifficulty();
    }

    // Thiết lập thông số dựa vào độ khó
    setupDifficulty() {
        switch (this.difficulty) {
            case 'easy':
                this.reactionTime = 1000; // 1 giây
                this.errorMargin = 40;    // Lỗi dự đoán lớn
                this.predictionSkill = 0.6; // Khả năng dự đoán kém
                this.maxSpeed = 3;        // Di chuyển chậm
                break;

            case 'medium':
                this.reactionTime = 700;  // 0.7 giây
                this.errorMargin = 20;
                this.predictionSkill = 0.8;
                this.maxSpeed = 5;
                break;

            case 'hard':
                this.reactionTime = 500;  // 0.5 giây
                this.errorMargin = 10;    // Lỗi dự đoán nhỏ
                this.predictionSkill = 0.95; // Khả năng dự đoán gần như hoàn hảo
                this.maxSpeed = 7;        // Di chuyển nhanh
                break;

            default:
                this.reactionTime = 700;
                this.errorMargin = 20;
                this.predictionSkill = 0.8;
                this.maxSpeed = 5;
        }
    }

    // Cập nhật vị trí paddle của AI
    update(ballX, ballY, ballSpeedX, ballSpeedY, currentTime) {
        // Chỉ cập nhật khi đủ thời gian phản ứng
        if (currentTime - this.lastUpdateTime < this.reactionTime) {
            return this.paddleY;
        }

        // Cập nhật thời gian
        this.lastUpdateTime = currentTime;

        // Thêm thông tin bóng vào lịch sử để phân tích
        this.ballHistory.push({x: ballX, y: ballY, dx: ballSpeedX, dy: ballSpeedY});
        if (this.ballHistory.length > this.maxHistory) {
            this.ballHistory.shift();
        }

        // Dự đoán vị trí bóng khi chạm đến paddle của AI
        if (ballSpeedX > 0) { // Bóng đang di chuyển về phía AI (phải)
            this.predictBallPosition(ballX, ballY, ballSpeedX, ballSpeedY);
        }

        // Di chuyển paddle dựa vào dự đoán
        if (this.predictedY !== null) {
            // Thêm lỗi dự đoán để tăng tính người thật
            const error = (Math.random() * 2 - 1) * this.errorMargin;
            const targetY = this.predictedY + error;

            // Giới hạn trong khoảng hợp lệ
            this.targetY = Math.max(this.paddleHeight / 2,
                          Math.min(this.fieldHeight - this.paddleHeight / 2, targetY));
        }

        // Di chuyển paddle với tốc độ giới hạn
        const paddleCenter = this.paddleY + this.paddleHeight / 2;
        const distance = this.targetY - paddleCenter;

        if (Math.abs(distance) > this.maxSpeed) {
            // Di chuyển với tốc độ tối đa
            this.paddleY += Math.sign(distance) * this.maxSpeed;
        } else {
            // Di chuyển trực tiếp đến vị trí mục tiêu
            this.paddleY += distance;
        }

        // Giới hạn paddle trong khoảng hợp lệ
        this.paddleY = Math.max(0, Math.min(this.fieldHeight - this.paddleHeight, this.paddleY));

        return this.paddleY;
    }

    // Dự đoán vị trí bóng khi chạm đến paddle phải
    predictBallPosition(ballX, ballY, ballSpeedX, ballSpeedY) {
        if (ballSpeedX <= 0) return; // Bóng đang di chuyển về phía khác

        // Tính toán thời gian để bóng đến paddle phải
        const rightSideX = 800 - 20 - this.paddleHeight/2; // Vị trí paddle phải
        const timeToReach = (rightSideX - ballX) / ballSpeedX;

        // Tính toán vị trí y của bóng tại thời điểm đó
        let predictedY = ballY + (ballSpeedY * timeToReach);

        // Xử lý các lần nảy trên tường
        const bounces = Math.floor(Math.abs(predictedY) / this.fieldHeight) +
                       Math.floor(Math.abs(predictedY - this.fieldHeight) / this.fieldHeight);

        if (bounces % 2 === 1) {
            predictedY = this.fieldHeight - Math.abs(predictedY % this.fieldHeight);
        } else {
            predictedY = Math.abs(predictedY % this.fieldHeight);
        }

        // Áp dụng độ chính xác dự đoán dựa vào độ khó
        if (this.predictionSkill < 1.0) {
            const skillError = (1.0 - this.predictionSkill) * this.fieldHeight * (Math.random() - 0.5);
            predictedY += skillError;
        }

        // Lưu giá trị dự đoán
        this.predictedY = predictedY;
    }

    // Đặt lại AI
    reset(initialY) {
        this.paddleY = initialY || this.fieldHeight / 2 - this.paddleHeight / 2;
        this.targetY = this.paddleY;
        this.predictedY = null;
        this.ballHistory = [];
        this.lastUpdateTime = 0;
    }
}